R/utility.R

```diff
@@ -725,4 +724,117 @@ viewSplines <- function(knots, degree, theta) {
+#' probability pairs that define the desired survival curve
+#' @return A vector of parameters that define the survival curve optimized for
+#' the target points. The first element of the vector represents the "f"
+#' parameter and the second element represents the "shape" parameter.
+#' @examples
+#' points <- list(c(60, 0.90), c(100, .75), c(200, .25), c(250, .10))
+#' survGetParams(points)
+#' @export
+#' @concept utility
+survGetParams <- function(points) {
+  
+  assertNotMissing(points = missing(points))
+  assertClass(points = points, class = "list")
+  
+  time <- sapply(points, function(x) x[1])
+  p <- sapply(points, function(x) x[2])
+  
+  assertPositive(time)
+  assertAscending(time)
+  
+  assertProbability(p)  
+  assertDescending(p)
+  
+  loss_surv <- function(params, points) {
+    
+    loss <- function(a, params) {
+      ( params[2]*(log(-log(a[2])) - params[1]) - log(a[1]) ) ^ 2
+    }
+    
+    sum(sapply(points, function(a) loss(a, params)))
+    
+  }
+  
+  r <- stats::optim(
+    par = c(1, 1), 
+    fn = loss_surv, 
+    points = points,
+    method = "L-BFGS-B", 
+    lower = c(-Inf, 0),
+    upper = c(Inf, Inf)
+  )
+  
+  if (r$convergence !=0) stop("Optimization did not converge")
+  
+  return(r$par)
+}
+
+#' Plot survival curves
+#' 
+#' @param f This is the "formula" parameter of the Weibull-based survival curve
+#' that can be used to define the scale of the distribution
+#' @param shape The parameter that defines the shape of the distribution.
+#' @param points An optional list of two-element vectors specifying the desired 
+#' time and probability pairs that define the desired survival curve. If no list
+#' is specified then the plot will not include any points.
+#' @param n The number of points along the curve that will be used to 
+#' define the line. Defaults to 100.
+#' @param scale An optional scale parameter that defaults to 1. If the value is 
+#' 1, the scale of the distribution is determined entirely by the argument "f".
+#' @return A ggplot of the survival curve defined by the specified parameters.
+#' If the argument points is specified, the plot will include them
+#' @examples
+#' points <- list(c(60, 0.90), c(100, .75), c(200, .25), c(250, .10))
+#' r <- survGetParams(points)
+#' survParamPlot(r[1], r[2])
+#' survParamPlot(r[1], r[2], points = points)
+#' @export
+#' @concept utility
+survParamPlot <- function(f, shape, points = NULL, n = 100, scale = 1) {
+  
+  assertNotMissing(f = missing(f), shape = missing(shape))
+  assertPositive(shape)
+  
+  # "declares" to avoid global NOTE
+  
+  V1 <- NULL
+  V2 <- NULL
+  
+  ###
+  
+  u <- seq(1, 0.001, length = n)
+  
+  dd <- data.table::data.table(
+    f = f,
+    scale = scale,
+    shape = shape,
+    T = (-(log(u)*scale/exp(f)))^(shape),
+    p = round(1 - cumsum(rep(1/length(u), length(u))), 3)
+  )
+  
+  p <- ggplot2::ggplot(data = dd, ggplot2::aes(x = T, y = p)) +
```

Comment:
As we only have `ggplot2` in `SUGGESTS` we would either need to add it as an `IMPORT` or guard against the usage (which I think is the preferred option as we only use it in 3 functions that are very non-essential) as suggested [here](https://r-pkgs.org/description.html#guarding-the-use-of-a-suggested-package). 

I just checked and `ViewSplines` and `ViewBasis` should also get the same treatment.

Reply:
Done.